modules/library/coverage/src/main/java/org/geotools/image/util/ImageUtilities.java

```diff
@@ -1361,18 +1361,74 @@ public static void setSubsamplingFactors(
+        // Ensures we never request a factor larger than the raster size
+        if (subSamplingFactorX > rasterWidth) subSamplingFactorX = rasterWidth;
+        if (subSamplingFactorY > rasterHeight) subSamplingFactorY = rasterHeight;
+
+        // Upward adjustment ONLY if it falls within tolerance
+        // Prevents unnecessary quality drops, but allows better alignment when very close
+        int candidateX = subSamplingFactorX + 1;
+        if (candidateX <= rasterWidth) {
+            double relErrFloor = Math.abs(ratioX - subSamplingFactorX) / Math.max(ratioX, 1e-9);
+            double relErrUp = Math.abs(ratioX - candidateX) / Math.max(ratioX, 1e-9);
+            if (relErrUp <= overviewTolerance && relErrUp < relErrFloor) {
+                subSamplingFactorX = candidateX;
+            }
+        }
+        int candidateY = subSamplingFactorY + 1;
+        if (candidateY <= rasterHeight) {
+            double relErrFloor = Math.abs(ratioY - subSamplingFactorY) / Math.max(ratioY, 1e-9);
+            double relErrUp = Math.abs(ratioY - candidateY) / Math.max(ratioY, 1e-9);
+            if (relErrUp <= overviewTolerance && relErrUp < relErrFloor) {
+                subSamplingFactorY = candidateY;
+            }
+        }
 
-        while (subSamplingFactorY > 0 && rasterHeight / subSamplingFactorY <= 0) subSamplingFactorY--;
-        subSamplingFactorY = subSamplingFactorY <= 0 ? 1 : subSamplingFactorY;
+        // Align SourceRegion to subsampling factors to avoid seams
+        // If there is a source region, adjust it so (x,y,width,height) are multiples of the factors
+        java.awt.Rectangle region = readParameters.getSourceRegion();
+        if (region != null) {
+            // Align x and y to the lower multiple of the factors
+            int alignedX = region.x - Math.floorMod(region.x, subSamplingFactorX);
+            int alignedY = region.y - Math.floorMod(region.y, subSamplingFactorY);
+
+            // Align width/height to the lower multiple of the factors (avoids half rows/columns)
+            int alignedW = region.width - Math.floorMod(region.width, subSamplingFactorX);
+            int alignedH = region.height - Math.floorMod(region.height, subSamplingFactorY);
+
+            // Ensure positive, non-zero sizes (at least one subsampling block)
+            if (alignedW <= 0) alignedW = subSamplingFactorX;
+            if (alignedH <= 0) alignedH = subSamplingFactorY;
+
+            // Conservative clipping if realignment exceeds raster bounds
+            // (assuming (0,0)-(rasterWidth,rasterHeight) as the full raster bounds)
+            if (alignedX < 0) alignedX = 0;
+            if (alignedY < 0) alignedY = 0;
+            if (alignedX + alignedW > rasterWidth) alignedW = Math.max(subSamplingFactorX, rasterWidth - alignedX);
+            if (alignedY + alignedH > rasterHeight) alignedH = Math.max(subSamplingFactorY, rasterHeight - alignedY);
```

Comment:
This still does not guarantee that we go above the raster width/height, could a reader throw an exception if we do, @dromagnoli ?